waku/v2/protocol/filter/client.go

```diff
@@ -379,7 +379,7 @@ func (wf *WakuFilterLightNode) Unsubscribe(ctx context.Context, contentFilter Co
 	localWg.Wait()
 	close(resultChan)
 	for _, peerID := range peersUnsubscribed {
-		if len(wf.subscriptions.items[peerID].subscriptionsPerTopic) == 0 {
+		if wf.subscriptions != nil && wf.subscriptions.items != nil && wf.subscriptions.items[peerID] != nil && len(wf.subscriptions.items[peerID].subscriptionsPerTopic) == 0 {
```

Comment:
Hm. In https://github.com/waku-org/go-waku/blob/6463dbeb709f5e5353a405f1117ae4d12c7f7c39/waku/v2/protocol/filter/filter_test.go#L224, right before the function ends, I added the following just to make sure if there was a problem due to calling Unsubscribe more than once:
```go
	_, err = s.lightNode.Unsubscribe(s.ctx, s.contentFilter, Peer(s.fullNodeHost.ID()))
	s.Require().NoError(err)
	_, err = s.lightNode.Unsubscribe(s.ctx, s.contentFilter, Peer(s.fullNodeHost.ID()))
	s.Require().NoError(err)
	_, err = s.lightNode.Unsubscribe(s.ctx, s.contentFilter, UnsubscribeAll())
	s.Require().NoError(err)
	_, err = s.lightNode.UnsubscribeAll(s.ctx, Peer(s.fullNodeHost.ID()))
	s.Require().NoError(err)
	_, err = s.lightNode.UnsubscribeAll(s.ctx, UnsubscribeAll())
	s.Require().NoError(err)
```
and i cannot generate the error. Perhaps is the unsubscribe being called after the node is stopped? if so, status-go should probably wait for the unsubscription to be completed before stopping the node?

Reply:
that's possible, though we can't wait for unsubscribe to finish if it's synchronous. 
That's not good for us, since if the user wants to logout, we don't want to have them wait, as we want login/logout to be rather quick, so unsubscribe should really be fire/forget ideally from the client perspective. 
In this case, i think the node should also make sure that it does not accept calls (some guard) when stopped (returning an error is fine, but panic in our case crashes the app)

Reply:
Do check this PR: https://github.com/waku-org/go-waku/pull/692
In here you'd call unsubscribe like this:

```
	_, err = s.lightNode.Unsubscribe(s.ctx, s.contentFilter, Peer(s.fullNodeHost.ID()), Async())
```
And that would do a 'fire and forget' for requesting unsubscriptions. I'll try in the same PR to add the guard to not accept subscriptions/unsubscriptions while filter is stopped

Reply:
Done! PR updated with the guard for checking the state of filter

Reply:
I would go for now with my changes, to err on the side of caution, it has been tested already by QA. 
The issue is the usage of `CompareAndSwap`, it only partially protect from race conditions, so in theory this could still happen (at least in this PR, and I'd like to minimize changes to get through as it's currently blocking some other changes). I can push a test to replicate in a separate PR and better explain the issue (you can't replicate in your tests because the node is started as part of the process, so it's already set up).

Reply:
Alright. let's merge the PR then!